consensus/istanbul/backend/announce.go

```diff
@@ -302,83 +390,92 @@ func (sb *Backend) handleIstAnnounce(payload []byte) error {
-		if !regAndActiveVals[announceRecord.DestAddress] || processedAddresses[announceRecord.DestAddress] {
-			msgHasDupsOrIrrelevantEntries = true
-			continue
+	if sb.coreStarted && regAndActiveVals[sb.Address()] {
+		var node *enode.Node
+		var processedAddresses = make(map[common.Address]bool)
+		for _, encryptedEnode := range announcePayload.EncryptedEnodes {
+			// Don't process duplicate entries or entries that are not in the regAndActive valset
+			if !regAndActiveVals[encryptedEnode.DecrypterAddress] || processedAddresses[encryptedEnode.DecrypterAddress] {
+				msgHasDupsOrIrrelevantEntries = true
+				continue
+			}
+
+			if encryptedEnode.DecrypterAddress == sb.Address() {
+				// TODO: Decrypt the enodeURL using this validator's validator key after making changes to encrypt it
+				enodeUrl := string(encryptedEnode.EncryptedEnodeURL)
+				node, err = enode.ParseV4(enodeUrl)
+				if err != nil {
+					logger.Error("Error in parsing enodeURL", "enodeUrl", enodeUrl)
+					return err
+				}
+			}
+			destAddresses = append(destAddresses, encryptedEnode.DecrypterAddress.String())
+			processedAddresses[encryptedEnode.DecrypterAddress] = true
 		}
 
-		if announceRecord.DestAddress == sb.Address() {
-			// TODO: Decrypt the enodeURL using this validator's validator key after making changes to encrypt it
-			enodeUrl := string(announceRecord.EncryptedEnodeURL)
-			node, err = enode.ParseV4(enodeUrl)
-			if err != nil {
-				logger.Error("Error in parsing enodeURL", "enodeUrl", enodeUrl)
+		// Save in the valEnodeTable
+		if node != nil {
+			if err := sb.valEnodeTable.Upsert(map[common.Address]*vet.AddressEntry{msg.Address: {Node: node, Timestamp: announcePayload.Timestamp}}); err != nil {
+				logger.Warn("Error in upserting a valenode entry", "AnnouncePayload", announcePayload.String(), "error", err)
 				return err
 			}
 		}
-		destAddresses = append(destAddresses, announceRecord.DestAddress.String())
-		processedAddresses[announceRecord.DestAddress] = true
-	}
-	// Save in the valEnodeTable if mining
-	if sb.coreStarted && node != nil {
-		if err := sb.valEnodeTable.Upsert(map[common.Address]*vet.AddressEntry{msg.Address: {Node: node, Timestamp: announceData.Timestamp}}); err != nil {
-			logger.Warn("Error in upserting a valenode entry", "AnnounceData", announceData.String(), "error", err)
-			return err
-		}
 	}
 
 	if !msgHasDupsOrIrrelevantEntries {
-		if err = sb.regossipIstAnnounce(msg, payload, announceData, regAndActiveVals, destAddresses); err != nil {
+		// Cache this announce message
+		sb.cachedAnnounceMsgs[msg.Address] = &announceMsgCachedEntry{MsgTimestamp: announcePayload.Timestamp,
+			MsgPayload: payload}
+
+		// Prune the announce cache for entries that are not in the current registered/elected validator set
+		for cachedValAddress := range sb.cachedAnnounceMsgs {
+			if !regAndActiveVals[cachedValAddress] {
+				delete(sb.cachedAnnounceMsgs, cachedValAddress)
+			}
+		}
```

Comment:
Why is this skipped when the message has duplicate or irrelevant entries. Also, although it is fine to handle this here, it also seems like something that could be handled out-of-band. I would suggest adding a comment explaining that this is here for a particular reason, or that it is arbitrary that it be handled here.

Reply:
I added a function `validateAnnounce` that will explain the reasononing.

Basically, we want to force the validators to construct as succint announce message as possible to minimize any network usage.